### DISCUSSION TITLE
Packaging components

### DISCUSSION THREAD

**Discussion by mikko-ahonen** [django-components/django-components#557] [Q&A] [answered]
I want to share code I have developed for Django. The typical method of sharing with Django is an app that can be included in the project.

The app I have been developing contains both regular views as well as django components used by these views.

How would you package it so it would be convenient for users? Is there a way for example to put components inside an app?

**Comment by JuroOravec** [comment, marked as answer]
Hey @mikko-ahonen, coincidentally, I was doing that just yesterday and today! Have a look at https://github.com/JuroOravec/alpinui/tree/main/packages/alpinui-django.

How to:
1. I defined the components inside the `<my_app>/templates` directory.
2. All components are "single-file components" (SFC), meaning that I've inlined the HTML and CSS files, so there are no non-python dependencies.
3. I imported the components files inside the `ready()` hook in `apps.py`, so the components are registered when user install the app with `INSTALLED_APPS`.
4. I've built the distribution wheel using the same command as used [in django-components](https://github.com/EmilStenstrom/django-components/blob/26a3f1473a38fdd15f957f84ebab7a5db7970999/.github/workflows/publish-to-pypi.yml#L32):
	```sh
	python -m build --sdist --wheel --outdir dist/ .
	```
5. I used `twine` to publish to PyPI ([See Python user guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/#uploading-the-distribution-archives))
	```sh
	twine upload --repository pypi dist/* -u __token__ -p <PyPI_TOKEN>
	```

Notes:
- I haven't tried how things would work if I defined the components inside a separate `components` directory.
- The user of the package NEEDS to have installed and configured `django_components`.
- If you have the HTML/CSS/JS files separate, you may need to define `MANIFEST.in` to include those files with the distribution ([see user guide](https://setuptools.pypa.io/en/latest/userguide/miscellaneous.html)). But I don't know if or how to then load those static files into Django.

**Comment by mikko-ahonen** [comment]
Thank you!
